sssd_test_framework/utils/ipa_ca.py

```diff
@@ -0,0 +1,236 @@
+from __future__ import annotations
+
+import os
+import re
+import shlex
+import uuid
+from typing import Dict, Literal, Optional, Tuple
+
+from pytest_mh import MultihostHost
+from pytest_mh.cli import CLIBuilder, CLIBuilderArgs
+from pytest_mh.utils.fs import LinuxFileSystem
+
+RevocationReason = Literal[
+    "unspecified",
+    "key_compromise",
+    "ca_compromise",
+    "affiliation_changed",
+    "superseded",
+    "cessation_of_operation",
+    "certificate_hold",
+    "remove_from_crl",
+    "privilege_withdrawn",
+    "aa_compromise",
+]
+
+
+class IPACertificateAuthority:
+    """
+    Provides helper methods for FreeIPA Certificate Authority operations.
+
+    This class allows requesting, revoking, placing/removing certificate holds,
+    and retrieving certificate information via the `ipa` CLI.
+    """
+
+    def __init__(self, host: MultihostHost, fs: LinuxFileSystem) -> None:
+        """
+        Initialize the IPA Certificate Authority helper.
+
+        Args:
+            host: A connected `MultihostHost` instance for running commands.
+            fs: A `LinuxFileSystem` instance for file operations.
+        """
+        self.host = host
+        self.fs = fs
+        self.cli: CLIBuilder = host.cli
+        self.temp_dir = f"/tmp/ipa_test_certs_{os.getpid()}_{uuid.uuid4().hex}"
+        self.fs.mkdir_p(self.temp_dir, mode="700")
+
+    def __del__(self) -> None:
+        """Clean up temporary files on object destruction."""
+        self.teardown()
+
+    def teardown(self) -> None:
+        """
+        Remove the temporary directory used for storing CSR, keys, and certificates.
+
+        Logs a warning if cleanup fails.
+        """
+        try:
+            self.fs.rm(self.temp_dir)
+        except Exception as e:
+            self.host.logger.warning(f"Failed to remove temporary directory {self.temp_dir}: {e}")
+
+    def request(
+        self,
+        principal: str,
+        subject: Optional[str] = None,
+        add_service: bool = False,
+        key_size: int = 2048,
+    ) -> Tuple[str, str, str]:
+        """
+        Request a certificate from the IPA CA.
+
+        Args:
+            principal: The principal (user or service) name.
+            subject: Optional OpenSSL subject (e.g., `/CN=example`). If omitted, derived from principal.
+            add_service: Whether to add the principal as an IPA service.
+            key_size: RSA key size in bits.
+
+        Returns:
+            A tuple of (certificate_path, key_path, csr_path).
+
+        Raises:
+            ValueError: If subject cannot be derived from principal.
+            RuntimeError: If CSR generation fails.
```

Comment:
I think we're supposed to list exceptions to raise like this:
```
    :raises ValueError: If subject cannot be derived from principal.
    :raises RuntimeError.: If CSR generation fails.
```
I think you can see examples of this in utils/sssd.py